cpp/include/cudf/rolling.hpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace experimental {
+
+/** 
+ * @brief Rolling window aggregation operations
+ */
+enum class rolling_operator {
+  SUM,       ///< Computes the sum of all values in the window
+  MIN,       ///< Computes minimum value in the window
+  MAX,       ///< Computes maximum value in the window
+  MEAN,       ///< Computes arithmetic mean of all values in the window
+  COUNT,     ///< Computes the number of values in the window
+  NUMBA_UDF, ///< A user-defined aggregation operation defined in PTX code generated by `numba`
+  CUDA_UDF,  ///< A generic aggregation operation defined in CUDA code
+  N_OPS,     ///< ALL NEW OPERATIONS SHOULD BE ADDED ABOVE THIS LINE
+};
+
+/**
+ * @brief  Applies a fixed-size rolling window function to the values in a column.
+ *
+ * This function aggregates values in a window around each element i of the input column, and
+ * invalidates the bit mask for element i if there are not enough observations. The window size is
+ * static (the same for each element). This matches Pandas' API for DataFrame.rolling with a few
+ * notable differences:
+ * - instead of the center flag it uses the forward window size to allow for more flexible windows.
+ *   The total window size = window + forward_window. Element i uses elements
+ *   [i-window+1, i+forward_window] to do the window computation.
```

Comment:
@nsakharnykh can you comment on this design choice? I'm afraid to change it lest we become incompatible with some driving use case.

Reply:
Pandas doesn't have `forward_window` parameter, instead they use `centered` to specify how the window is aligned with the current index. We went with `forward_window` because it give more flexibility (larger window to look back and smaller window to look forward?), and the decision to make `forward_window` inclusive / `window`. exclusive is based on gQuant's behavior (that we used as a starting point for the rolling windows design): https://github.com/rapidsai/gQuant/blob/f117268c3480a820ac8a93d90980192cec014883/gquant/cuindicator/rolling.py#L93. Looking at the SQL specs there are corresponding "preceding" and "following" parameters to control the window/forward_window sizes (https://www.red-gate.com/simple-talk/sql/learn-sql-server/window-functions-in-sql-server-part-2-the-frame/), and both are inclusive. I'm fine making the change to have both options to be inclusive, in fact, we could also rename the window sizes to `preceding_window_size`  and `following_window_size` to make it more clear.

Reply:
Thanks for the details! Will make these changes.